File: LC445.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        s1 = []
        s2 = []

        while l1:
            s1.append(l1.val)
            l1 = l1.next
        
        while l2:
            s2.append(l2.val)
            l2 = l2.next
        
        s = 0
        res = ListNode(0)

        while len(s1) != 0 or len(s2) != 0:
            if len(s1) != 0:
                s += s1.pop()
            
            if len(s2) != 0:
                s += s2.pop()
            
            res.val = s%10
            c = s//10
            t = ListNode(c)
            t.next = res
            res = t
            s = c
        
        return res.next if c == 0 else res

File: LC445.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        stack <int> s1, s2;

        while(l1){
            s1.push(l1->val);
            l1 = l1->next;
        }

        while(l2){
            s2.push(l2->val);
            l2 = l2->next;
        }

        int sum = 0, c = 0;
        ListNode* res = new ListNode();
        while(!s1.empty() || !s2.empty()){

            if(!s1.empty()){
                sum += s1.top();
                s1.pop();
            }

            if(!s2.empty()){
                sum += s2.top();
                s2.pop();
            }

            res->val = sum % 10;
            c = sum / 10;
            ListNode* t = new ListNode(c);
            t->next = res;
            res = t;
            sum = c;
        }

        return c == 0? res->next : res;
    }
};
